In [1]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

%opts magic unavailable (pyparsing cannot be imported)
%compositor magic unavailable (pyparsing cannot be imported)


In [2]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [3]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [4]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [5]:
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [6]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
from sklearn.preprocessing import StandardScaler

features = ['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y']

sc = StandardScaler()
sc_feature_arrays = sc.fit_transform(df_market_data[features])

In [7]:
df_market_data.index

Index(['bitcoin', 'ethereum', 'tether', 'ripple', 'bitcoin-cash',
       'binancecoin', 'chainlink', 'cardano', 'litecoin', 'bitcoin-cash-sv',
       'crypto-com-chain', 'usd-coin', 'eos', 'monero', 'tron', 'tezos', 'okb',
       'stellar', 'cosmos', 'cdai', 'neo', 'wrapped-bitcoin', 'leo-token',
       'huobi-token', 'nem', 'binance-usd', 'iota', 'vechain', 'zcash',
       'theta-token', 'dash', 'ethereum-classic', 'ethlend', 'maker', 'havven',
       'omisego', 'celsius-degree-token', 'ontology', 'ftx-token', 'true-usd',
       'digibyte'],
      dtype='object', name='coin_id')

In [8]:
# Create a DataFrame with the scaled data
# Copy the crypto names from the original DataFrame

df_market_data_scaled = pd.DataFrame(sc_feature_arrays,columns=features)

df_market_data_scaled.reset_index()
# Set the coin_id column as index
df_market_data_scaled.set_index(df_market_data.index,inplace=True)

# Display the scaled DataFrame
df_market_data_scaled.head(50)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

---

### Find the Best Value for k Using the Original Data.

In [9]:
# Create a list with the number of k-values from 1 to 11
no_cluster_list = [i for i in range(1,12)]
no_cluster_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [10]:
# Create an empty list to store the inertia values
inertia_values_list = []

# Create a for loop to compute the inertia with each possible value of k

for test_k in no_cluster_list:
    # Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_means_model = KMeans(n_clusters=test_k,init="k-means++",random_state=0)
# 2. Fit the model to the data using `df_market_data_scaled`
    k_means_model.fit(df_market_data_scaled)
# 3. Append the model.inertia_ to the inertia list
    inertia_values_list.append(k_means_model.inertia_)




In [11]:
inertia_values_list

[287.0,
 198.57181809798126,
 145.63574772192862,
 79.02243535120975,
 66.21778006000338,
 59.13907002537009,
 48.406314490413614,
 38.99358798927969,
 34.40059545400235,
 28.57360256096785,
 23.890044355377952]

In [12]:
# Create a dictionary with the data to plot the Elbow curve
number_cluster_inertia_value_dict = { 'number_of_cluster':no_cluster_list,
                                     'inertia_based_on_no_cluster':inertia_values_list
                                    }

# Create a DataFrame with the data to plot the Elbow curve
df_cluster_data = pd.DataFrame(number_cluster_inertia_value_dict)
df_cluster_data

number_of_cluster  inertia_based_on_no_cluster
0                   1                   287.000000
1                   2                   198.571818
2                   3                   145.635748
3                   4                    79.022435
4                   5                    66.217780
5                   6                    59.139070
6                   7                    48.406314
7                   8                    38.993588
8                   9                    34.400595
9                  10                    28.573603
10                 11                    23.890044

In [13]:
# Plot a line chart with all the inertia values computed with

# the different values of k to visually identify the optimal value for k.


elbow_curve_origional_df = df_cluster_data.hvplot.line(
    x='number_of_cluster',
    y='inertia_based_on_no_cluster',
    title="Elbow Curve For Original Data",
    width=400,
    height=300
)
elbow_curve_origional_df

:Curve   [number_of_cluster]   (inertia_based_on_no_cluster)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** We have to select the value of k at the “elbow” that is the point after which the distortion/inertia starts decreasing in a linear fashion. Thus for the given data, we conclude that the optimal number of clusters for the data is 4.

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

---

In [14]:
# Initialise the K-Means model using the best value for k
k_means_model = KMeans(n_clusters=4,init="k-means++",random_state=0)

In [15]:
# Fit the K-Means model using the scaled DataFrame
k_means_model.fit(df_market_data_scaled)

KMeans(n_clusters=4, random_state=0)

In [16]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
array_cluster_value = k_means_model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
print(array_cluster_value)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [17]:
# Create a copy of the scaled DataFrame
copy_df_market_data_scaled = df_market_data_scaled.copy()

In [18]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
copy_df_market_data_scaled['predicted_cluster'] = array_cluster_value

# Display the copy of the scaled DataFrame
copy_df_market_data_scaled.head(50)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                         
bitcoin                                  0.508529                    0.493193   
ethereum                                 0.185446                    0.934445   
tether                                   0.021774                   -0.706337   
ripple                                  -0.040764                   -0.810928   
bitcoin-cash                             1.193036                    2.000959   
binancecoin                              0.891871                    1.327295   
chainlink                                0.011397                    2.572251   
cardano                                  0.102530                    1.508001   
litecoin                                 0.077497                    0.334297   
bitcoin-cash-sv                          0.448952                   -0.190684   
crypto-com-chain                         0.331280                   -1.614844   
usd-coin                                 0.034352                   -0.733026   
eos                                      0.155710                   -0.922491   
monero                                   0.262723                    1.792602   
tron                                     0.130050                   -0.041018   
tezos                                   -0.151583                    0.708196   
okb                                     -0.923203                   -1.437359   
stellar                                 -0.277543                   -0.385209   
cosmos                                  -0.255978                    1.840274   
cdai                                     0.180851                   -0.704931   
neo                                      0.286546                   -0.326301   
wrapped-bitcoin                          0.515453                    0.461843   
leo-token                                0.051758                   -0.928381   
huobi-token                             -0.052032                   -0.457229   
nem                                     -0.217984                   -0.849381   
binance-usd                              0.061339                   -0.706669   
iota                                     0.259097                    0.249508   
vechain                                  0.585089                   -0.994231   
zcash                                   -0.127467                    0.929119   
theta-token                             -1.612188                   -1.682027   
dash                                    -0.296940                    0.094763   
ethereum-classic                        -0.071312                   -0.229484   
ethlend                                 -4.981042                   -0.045178   
maker                                   -0.125168                    0.580730   
havven                                  -1.428574                   -0.025510   
omisego                                  1.919812                    0.370447   
celsius-degree-token                     1.045530                   -0.618328   
ontology                                -0.409044                   -0.906963   
ftx-token                                0.414711                    0.414044   
true-usd                                 0.078038                   -0.687745   
digibyte                                 1.217453                   -0.607714   

                      price_change_percentage_14d  \
coin_id                                             
bitcoin                                  0.772200   
ethereum                                 0.558692   
tether                                  -0.021680   
ripple                                   0.249458   
bitcoin-cash                             1.760610   
binancecoin                              0.800214   
chainlink                                1.101647   
cardano                                  0.648885   
litecoin

In [19]:
copy_df_market_data_scaled.reset_index(inplace=True)

In [20]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Colour the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
x="price_change_percentage_24h"
y="price_change_percentage_7d"
scatter_plot_orginal_scaled_df = copy_df_market_data_scaled.hvplot.scatter(x=x, y=y,
                  legend='top',height=300,width=400,by='predicted_cluster', 
                  hover_cols=['coin_id','predicted_cluster'],title="Scatter Plot Original Data")

scatter_plot_orginal_scaled_df

:NdOverlay   [predicted_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id,predicted_cluster)

---

### Optimise Clusters with Principal Component Analysis.

In [21]:
# Create a PCA model instance and set `n_components=3`.
from sklearn.decomposition import PCA
pca = PCA(n_components=3)

In [22]:
df_market_data_scaled.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [23]:
features

['price_change_percentage_24h',
 'price_change_percentage_7d',
 'price_change_percentage_14d',
 'price_change_percentage_30d',
 'price_change_percentage_60d',
 'price_change_percentage_200d',
 'price_change_percentage_1y']

In [24]:
# Use the PCA model with `fit_transform` to reduce the original scaled Dataframe
# down to three principal components.
pca_3_component = pca.fit_transform(df_market_data_scaled)

# View the scaled PCA data
pca_3_component

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527],
       [-0.51653377,  1.38837748,  0.80407131],
       [-0.45071134,  0.51769912,  2.84614316],
       [-0.34559977,  0.72943939,  1.47801284],
       [-0.64946792,  0.43216514,  0.60030286],
       [-0.75901394, -0.20119979, -0.21765292],
       [-0.24819846, -1.37625159, -1.46202571],
       [-0.43840762, -0.17533654, -0.6633884 ],
       [-0.69342533, -0.47381462, -0.52759693],
       [ 0.06049915,  2.90940385,  1.49857131],
       [-0.39335243, -0.10819197, -0.01275608],
       [-0.79617564, -0.49440875,  1.08281169],
       [ 0.06407452, -1.26982514, -1.09882928],
       [-0.48901506, -0.73271912, -0.06254323],
       [-0.3062723 ,  0.70341515,  1.71422359],
       [-0.51352775, -0.14280239, -0.65656583],
       [-0.36212044, -0.98691441, -0.728

In [25]:
# Retrieve the explained variance to determine how much information
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [26]:
sum(pca.explained_variance_ratio_)

np.float64(0.8950316570309842)

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is 0.895.

In [27]:
# Create a new DataFrame with the PCA data.
column_name = ["PC1","PC2","PC3"]
df_market_data_pca = pd.DataFrame(pca_3_component,columns=column_name)



# Copy the crypto names from the original scaled DataFrame
crypto_name = df_market_data_scaled.index
# Set the coin_id column as index
df_market_data_pca.set_index(crypto_name,inplace=True)


# Display the scaled PCA DataFrame
df_market_data_pca.head(10)

PC1       PC2       PC3
coin_id                                      
bitcoin         -0.600667  0.842760  0.461595
ethereum        -0.458261  0.458466  0.952877
tether          -0.433070 -0.168126 -0.641752
ripple          -0.471835 -0.222660 -0.479053
bitcoin-cash    -1.157800  2.041209  1.859715
binancecoin     -0.516534  1.388377  0.804071
chainlink       -0.450711  0.517699  2.846143
cardano         -0.345600  0.729439  1.478013
litecoin        -0.649468  0.432165  0.600303
bitcoin-cash-sv -0.759014 -0.201200 -0.217653

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [28]:
# Create a list with the number of k-values from 1 to 11
number_of_cluster_list = [i for i in range(1,12)]
number_of_cluster_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [29]:
# Create an empty list to store the inertia values
inertia_values_list_pca = []

# Create a for loop to compute the inertia with each possible value of k

# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for clusters in number_of_cluster_list:
    # Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_means_model = KMeans(n_clusters=clusters,init="k-means++",random_state=0)
# 2. Fit the model to the data using `df_market_data_scaled`
    k_means_model.fit(df_market_data_pca)
# 3. Append the model.inertia_ to the inertia list
    inertia_values_list_pca.append(k_means_model.inertia_)
inertia_values_list_pca


[256.87408556789245,
 168.81189630685213,
 115.91807663320061,
 49.66549665179737,
 41.357313803128555,
 30.796641787041587,
 25.91545691045613,
 21.890936274633226,
 17.720698854473664,
 13.83869870886457,
 7.984354811896226]

In [30]:
# Create a dictionary with the data to plot the Elbow curve
number_cluster_inertia_value_dict_pca = { 'number_of_cluster':number_of_cluster_list,
                                     'inertia_based_on_no_cluster':inertia_values_list_pca
                                    }

# Create a DataFrame with the data to plot the Elbow curve
df_cluster_data_pca = pd.DataFrame(number_cluster_inertia_value_dict_pca)
df_cluster_data_pca


number_of_cluster  inertia_based_on_no_cluster
0                   1                   256.874086
1                   2                   168.811896
2                   3                   115.918077
3                   4                    49.665497
4                   5                    41.357314
5                   6                    30.796642
6                   7                    25.915457
7                   8                    21.890936
8                   9                    17.720699
9                  10                    13.838699
10                 11                     7.984355

In [31]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
elbow_curve_pca_df = df_cluster_data_pca.hvplot.line(
    x='number_of_cluster',
    y='inertia_based_on_no_cluster',
    title="Elbow Curve After PCA",
    width=400,
    height=300
)
elbow_curve_pca_df



:Curve   [number_of_cluster]   (inertia_based_on_no_cluster)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** we have to select the value of k at the “elbow” that the point after which the distortion/inertia starts decreasing in a linear fashion. Thus for the given data, we conclude that the optimal number of clusters for the data is 4.


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** It is the same as previous that is k value 4 because “elbow” is the point after which the distortion/inertia starts decreasing in a linear fashion after 4 only.

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [32]:
# Initialise the K-Means model using the best value for k
k_means_model = KMeans(n_clusters=4,init="k-means++",random_state=0)

In [33]:
# Fit the K-Means model using the scaled PCA DataFrame
# Fit the K-Means model using the scaled DataFrame
k_means_model.fit(df_market_data_pca)

KMeans(n_clusters=4, random_state=0)

In [34]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
array_cluster_value = k_means_model.predict(df_market_data_pca)


# Print the resulting array of cluster values.
print(array_cluster_value)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 1 2 0 0 3
 0 0 0 0]


In [35]:
# Create a copy of the scaled PCA DataFrame
copy_df_market_data_scaled_pca = df_market_data_pca.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
copy_df_market_data_scaled_pca['predicted_cluster'] = array_cluster_value


# Display the copy of the scaled PCA DataFrame
copy_df_market_data_scaled_pca.head()

PC1       PC2       PC3  predicted_cluster
coin_id                                                      
bitcoin      -0.600667  0.842760  0.461595                  2
ethereum     -0.458261  0.458466  0.952877                  2
tether       -0.433070 -0.168126 -0.641752                  0
ripple       -0.471835 -0.222660 -0.479053                  0
bitcoin-cash -1.157800  2.041209  1.859715                  2

In [36]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Colour the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
x="PC1"
y="PC2"
#copy_df_market_data_scaled_pca.reset_index(inplace=True)
scatter_plot_pca_scaled_df = copy_df_market_data_scaled_pca.hvplot.scatter(x=x, y=y,
                  legend='top',by='predicted_cluster',width=400,height=300,
                  hover_cols=['coin_id','predicted_cluster'],title="Scatter Plot PCA")
scatter_plot_pca_scaled_df

:NdOverlay   [predicted_cluster]
   :Scatter   [PC1]   (PC2,coin_id,predicted_cluster)

### Visualise and Compare the Results

In this section, you will visually analyse the cluster analysis results by contrasting the outcome with and without using the optimisation techniques.

In [37]:
# Composite plot to contrast the Elbow curves

# YOUR CODE HERE!

elbow_curve_origional_df + elbow_curve_pca_df

:Layout
   .Curve.I  :Curve   [number_of_cluster]   (inertia_based_on_no_cluster)
   .Curve.II :Curve   [number_of_cluster]   (inertia_based_on_no_cluster)

In [38]:
# Composite plot to contrast the clusters
# YOUR CODE HERE!

scatter_plot_orginal_scaled_df + scatter_plot_pca_scaled_df

:Layout
   .NdOverlay.I  :NdOverlay   [predicted_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id,predicted_cluster)
   .NdOverlay.II :NdOverlay   [predicted_cluster]
      :Scatter   [PC1]   (PC2,coin_id,predicted_cluster)

#### Answer the following question: 

  * **Question:** After visually analysing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** After visually analysing the cluster analysis results, when less variables are used to cluster the data, the biggest clusters will tend to be much closer to each other compared to the original data. The PCA plot, which utilizes less variables, has also separated the cluster 3 (coin_id = "celsius-degree-token") point from cluster 0 and 2 which is not the case in the original data plot where cluster 3 (coin_id = "celsius-degree-token") is inside cluster 0. It perceivably indicates to me that in the original data, the coin_id “celsius_degree_token” has a certain feature standing out from the rest that is forming another cluster (a least in some other dimension) except that this extra characteristic will not manifest in that plot. In this regard, I find the PCA visualize data helpful in explaining the characteristics of the clusters and this particularly about the number of clusters is perfectly ok.